content/general/concepts/abstract-factory/abstract-factory.md

```diff
@@ -0,0 +1,381 @@
+    this.dateOpened = dateOpened;
+    this.balance = initBalance;
+  }
+
+  public void addFunds(final int amount) {
+    balance += amount;
+  }
+
+  public void withdrawFunds(final int amount) {
+    balance -= amount;
+  }
+  // Enforces concrete classes to override this method
+  public abstract void addInterest();
+
+  // Getters and setter
+}
+```
+
+Much like `CurrentAccount`, `SavingAccount` provides the common fields and constructor and requires its children to implement the `addInterest` method. This is the differentiating feature described between each savings account.
+
+Below provides the concrete implementations of `SavingAccount` for this example:
+
+```java
+public class GoldSaver extends SavingAccount {
+  // The Gold Saver has a high interest rate
+  public static final double INTEREST_RATE_MULTIPLIER = 1.05;
+
+  public GoldSaver(Customer accountHolder, LocalDate dateOpened, double initBalance) {
+    super(accountHolder, dateOpened, initBalance);
+  }
+  // Concrete overridden method
+  @Override
+  public void addInterest() {
+    if (LocalDate.now().getMonth() == this.getDateOpened().getMonth()
+            && LocalDate.now().getDayOfMonth() == this.getDateOpened().getDayOfMonth()) {
+      this.setBalance(this.getBalance() * INTEREST_RATE_MULTIPLIER);
+    }
+  }
+}
+```
+
+```java
+public class SilverSaver extends SavingAccount {
+  // The Silver Saver has an mid interest rate
+  public static final double INTEREST_RATE_MULTIPLIER = 1.03;
+
+  public SilverSaver(Customer accountHolder, LocalDate dateOpened, double initBalance) {
+    super(accountHolder, dateOpened, initBalance);
+  }
+  // Concrete overridden method
+  @Override
+  public void addInterest() {
+    if (LocalDate.now().getMonth() == this.getDateOpened().getMonth()
+            && LocalDate.now().getDayOfMonth() == this.getDateOpened().getDayOfMonth()) {
+      this.setBalance(this.getBalance() * INTEREST_RATE_MULTIPLIER);
+    }
+  }
+}
+```
+
+```java
+public class BronzeSaver extends SavingAccount {
+  // The Bronze Saver has a low interest rate
+  public static final double INTEREST_RATE_MULTIPLIER = 1.015;
+
+  public BronzeSaver(Customer accountHolder, LocalDate dateOpened, int initBalance) {
+        super(accountHolder, dateOpened, initBalance);
+  }
+  // Concrete overridden method
+  @Override
+  public void addInterest() {
+    if (LocalDate.now().getMonth() == this.getDateOpened().getMonth()
+            && LocalDate.now().getDayOfMonth() == this.getDateOpened().getDayOfMonth()) {
+      this.setBalance(this.getBalance() * INTEREST_RATE_MULTIPLIER);
+    }
+  }
+}
+```
+
+In each concrete savings account, a `static` variable (`INTEREST_RATE_MULTIPLIER`) has been provided and constructed with the correct value. This is then used to calculate interest in the `addInterest` method.
+
+The purpose of the `AccountFactory` is to return one of its concrete factories. Below provides the example of the abstract factory `AccountFactory`.
```

Comment:
```suggestion
The purpose of the abstract `AccountFactory` is to return one of its concrete factories. Below is an example of what the code would look like:
```

Per the code below, does the `<T>` mean in `AccountFactory<T>`?

Reply:
The <T> refers to something called Generics in java. Unlike languages like js and python, Java is strongly typed. This means that properties and variables have a type (either complex or simple) assigned. For e.g.

public **String** someString = "someString"; <- this property cant be anything other than a string.

The issue with that is when inheritance starts coming in, we sometimes don't know what type is needed until we are in a class that is lower down. Being able to pass children an unknown type, then in the child defining that type becomes really useful. You can see more here https://www.youtube.com/watch?v=h7piyWnQbZA 

The only other thing worth saying is what does <?> mean. It is literally shorthand for extends object, but the way java works is every type ultimately extends object. So the programmer is kind of saying 'I'm too lazy to work out what should go in here'. Its a little hacky.

Reply:
Nice! I just checked the Docs site, in the Java section, and we don't seem to have an entry for generic types. I'll make an issue for that one! No need to explain it here, though.